src/uffd_handler/src/prefaulter.rs

```diff
@@ -21,11 +26,266 @@ impl UffdPrefaulter {
+                prefault_start_addr += page_size;
+                // Found the first Uninitialized occurrence, so prefaulting will start from there.
+                // This unwrap is safe because it is guaranteed that the key is the start of a page
+                // within the current region.
+                if *region.page_states.get(&prefault_start_addr).unwrap()
+                    == MemPageState::Uninitialized
+                {
+                    break;
+                }
+            }
+        }
+
+        Ok(prefault_start_addr)
+    }
+
+    /// Return the end address of the chunk to prefault into memory. The address is obtained
+    /// by adding `after` amount of bytes to the end of the faulting page. If the end address
+    /// exceeds the current region, the chunk is truncated to fit into the memory region
+    /// (the end of the current memory region is returned).
+    /// When there is no customized prefaulting configured, the minimum chunk to copy to RAM
+    /// is the faulting page.
+    pub fn get_prefaulting_end_address(
+        &self,
+        page_fault_start_addr: usize,
+        region: &MemRegion,
+    ) -> Result<usize, HandlerError> {
+        let page_size = get_page_size().map_err(HandlerError::PageSize)?;
+
+        // This will not overflow because it has been verified beforehand during mappings
+        // deserialization.
+        let region_end = region.mapping.base_host_virt_addr + region.mapping.size;
+        // We add `page_size` because we always want to at least bring into RAM the faulting page,
+        // even if customized prefaulting is not configured.
+        let prefault_chunk_end = self
+            .bytes_after
+            .checked_add(page_fault_start_addr + page_size)
+            .unwrap_or(region_end);
+
+        // Make sure that the end of the chunk to prefault is within the current region.
+        Ok(min(prefault_chunk_end, region_end))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use utils::GuestRegionUffdMapping;
+
+    use super::*;
+    use crate::memory_region::create_mem_regions;
+
+    fn setup_region() -> Vec<MemRegion> {
+        let page_size = get_page_size().unwrap();
+        // Create a memory region with size 5 * page_size starting from 0x0 address.
+        let mappings = vec![GuestRegionUffdMapping {
+            base_host_virt_addr: 0x0,
+            size: page_size * 5,
+            offset: 0,
+        }];
+
+        create_mem_regions(mappings)
+    }
+
+    #[test]
+    fn test_init_prefaulter() {
+        let prefaulter = UffdPrefaulter::new(512, 1024).unwrap();
+        assert_eq!(prefaulter.bytes_before, 512 * KIB_MULT);
+        assert_eq!(prefaulter.bytes_after, 1024 * KIB_MULT);
+
+        let res = UffdPrefaulter::new(usize::MAX, 4);
+        assert!(res.is_err());
+        assert_eq!(
+            res.err().unwrap().to_string(),
+            PrefaulterError::InvalidAmount("before".to_string(), usize::MAX).to_string()
```

Comment:
nit: Can we have these `"before"`/`"after"` strings either be `&'static str`, or even better imo, distinct enum variants?